### DISCUSSION TITLE
Using a generic without a schema

### DISCUSSION THREAD

**Discussion by DaSchTour** [cyrilletuzi/angular-async-local-storage#998] [Q&A] [answered]
# Checks before posting an issue

- [x] I have read _all_ the library's documentation:
  - [README](https://github.com/cyrilletuzi/angular-async-local-storage/blob/main/README.md)
  - [CHANGELOG](https://github.com/cyrilletuzi/angular-async-local-storage/blob/main/CHANGELOG.md)
  - [MIGRATION](https://github.com/cyrilletuzi/angular-async-local-storage/blob/main/MIGRATION.md)
  - [/docs](https://github.com/cyrilletuzi/angular-async-local-storage/tree/main/docs)
- [x] I have checked in already:
  - [opened issues](https://github.com/cyrilletuzi/angular-async-local-storage/issues)
  - [closed issues](https://github.com/cyrilletuzi/angular-async-local-storage/issues?q=is%3Aissue+is%3Aclosed)
  - [questions answered](https://github.com/cyrilletuzi/angular-async-local-storage/discussions/categories/q-a) in the new "Discussions" tab
- [x] I understand I am writing to a *human being*, doing this open source project *unpaid*
- [x] I will be involved in my own issue, by answering questions and informing if the issue is resolved

# Description of the issue

The documentation states, that schema is optional, so the signature for generic watch should not force usage of schema.

So please change `watch<T>(index: string, schema: JSONSchema): Observable<T> {}` to `watch<T>(index: string, schema?: JSONSchema): Observable<T> {}`
This would allow to specify the type and avoid later typecasting.

**Comment by cyrilletuzi** [comment, marked as answer]
Hello,

The `watch()` signature is the same as the `get()` signature and is working as intended.

The `schema` *is* optional, but then an `Observable<unknown>` is returned, which is the intended behavior. Because if a *runtime* time check was not done with a `schema`, it is not right to rely on a *compilation* cast: indeed, it is *client-side* storage, so the data could have been forged, which could lead to bugs and security issues.

I invite you to read again the ["Validation" documentation](https://github.com/cyrilletuzi/angular-async-local-storage/blob/main/docs/VALIDATION.md), and especially the ["Why a schema and a cast?" section](https://github.com/cyrilletuzi/angular-async-local-storage/blob/main/docs/VALIDATION.md#why-a-schema-and-a-cast) for a longer description of the behavior.

Converting this to a discussion as it is not a bug.

**Reply by DaSchTour** [reply]
Yeah that's a valid point. IMHO I still would allow developers to use libraries at their intent and not force a certain opinion on them.
